go/ct/common/hash256.go

```diff
@@ -0,0 +1,116 @@
+//
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use
+// of this software will be governed by the GNU Lesser General Public Licence v3
+//
+
+package common
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/Fantom-foundation/Tosca/go/vm"
+	"pgregory.net/rand"
+)
+
+// Hash256 is an immutable array of Hash256 that can be trivially cloned.
+type Hash256 struct {
+	data *[256]vm.Hash
+}
+
+func NewHash256(hashes ...vm.Hash) Hash256 {
+	var data [256]vm.Hash
+	copy(data[:], hashes)
+	return Hash256{data: &data}
+}
+
+func (b Hash256) Equal(other Hash256) bool {
+	if b.data == other.data {
+		return true
+	}
+	if b.data == nil || other.data == nil {
+		return false
+	}
+	for i := 0; i < 256; i++ {
+		if b.data[i] != other.data[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func Randomhash256(rnd *rand.Rand) Hash256 {
+	hash256 := Hash256{}
+	hash256.data = new([256]vm.Hash)
+	for i := 0; i < 256; i++ {
+		hash256.data[i] = GetRandomHash(rnd)
+	}
+	return hash256
+}
+
+func (b Hash256) ToBytes() []byte {
+	bytes := []byte{}
+	for i := 0; i < 256; i++ {
+		bytes = append(bytes, b.data[i][:]...)
+	}
+	return bytes
+}
+
+func (b Hash256) FromBytes(bytes []byte) Hash256 {
+	hash256 := Hash256{}
+	for i := 0; i < 256; i++ {
+		copy(hash256.data[i][:], bytes[i*32:(i+1)*32])
+	}
+	return hash256
+}
+
+func (b Hash256) ToHash() [256]vm.Hash {
+	return [256]vm.Hash(*b.data)
+}
+
+func (b Hash256) String() string {
+	return fmt.Sprintf("%x", b.data)
+}
+
+func (b Hash256) MarshalJSON() ([]byte, error) {
+	return json.Marshal(b.data)
+}
+
+// Implement the UnmarshalJSON method for Hash256
+func (h *Hash256) UnmarshalJSON(data []byte) error {
+	// Unmarshal the JSON array into a slice of vm.Hash
+	var hashes [256]vm.Hash
+	err := json.Unmarshal(data, &hashes)
+	if err != nil {
+		return err
+	}
+
+	// Check the length of the Hash256 data
+	if len(hashes) != len(h.data) {
+		return fmt.Errorf("invalid Hash256 length")
+	}
+
+	// Copy the slice into the Hash256 array
+	h.data = &hashes
+	return nil
+}
+
+func (b Hash256) Get(index uint64) vm.Hash {
+	if index >= 256 {
+		panic("index out of range")
```

Comment:
Is this really a panic? Of course, the caller side cannot call this function out of range, but If we ever reuse this function somewhere else, the reviewer will not see this error, and we will hide the panic under a combination of runtime values. If instead you return an error and handle it in the adapter properly, this code will never panic.